src/renderers/dom/fiber/ReactDOMFiberEntry.js

```diff
@@ -601,6 +607,25 @@ function renderSubtreeIntoContainer(
         container.removeChild(rootSibling);
       }
     }
+    if (__DEV__) {
+      if (shouldHydrate && !forceHydrate && !warnedAboutHydrateAPI) {
+        warnedAboutHydrateAPI = true;
+        lowPriorityWarning(
+          false,
+          'render(): Calling ReactDOM.render() to hydrate server-rendered markup ' +
+            'will stop working in React v17. Replace the ReactDOM.render() call ' +
+            'with ReactDOM.hydrate() if you want React to attach to the server HTML.',
+        );
+      }
+      if (forceHydrate && !container.firstChild && !warnedAboutEmptyContainer) {
+        warnedAboutEmptyContainer = true;
+        warning(
+          false,
+          'hydrate(): Expected to hydrate from server-rendered markup, but the passed ' +
+            'DOM container node was empty. React will create the DOM from scratch.',
```

Comment:
IMO we should revert this warning since it's valid to have an empty container if your server rendered `""`, `null`, `false` or `[]` at the root. You'll get a warning about hydration being a non-match later. We can potentially special case the bulk error message in #10085 if the only diff was that the container was empty.